pandas/tests/reshape/test_concat.py

```diff
@@ -2249,3 +2295,109 @@ def test_concat_empty_and_non_empty_series_regression():
+
+    tm.assert_frame_equal(result, expected)
+
+
+def test_concat_sorts_index():
+    df1 = pd.DataFrame({"a": [1, 2, 3]}, index=['c', 'a', 'b'])
+    df2 = pd.DataFrame({"b": [1, 2]}, index=['a', 'b'])
+
+    with tm.assert_produces_warning(FutureWarning):
+        result = pd.concat([df1, df2], axis=1)
+
+    expected = pd.DataFrame({"a": [2, 3, 1], "b": [1, 2, None]},
+                            index=['a', 'b', 'c'],
+                            columns=['a', 'b'])
+    tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize('sort', [None, False, True])
+def test_concat_inner_sort(sort):
+    # https://github.com/pandas-dev/pandas/pull/20613
+    df1 = pd.DataFrame({"a": [1, 2], "b": [1, 2], "c": [1, 2]},
+                       columns=['b', 'a', 'c'])
+    df2 = pd.DataFrame({"a": [1, 2], 'b': [3, 4]}, index=[3, 4])
+
+    if sort is None:
+        with tm.assert_produces_warning(None):
+            # unset sort should *not* warn for inner join
+            # since that never sorted
+            result = pd.concat([df1, df2], sort=sort, join='inner',
+                               ignore_index=True)
+    else:
+        result = pd.concat([df1, df2], sort=sort, join='inner',
+                           ignore_index=True)
+
+    expected = pd.DataFrame({"b": [1, 2, 3, 4], "a": [1, 2, 1, 2]},
+                            columns=['b', 'a'])
+    if sort:
+        expected = expected[['a', 'b']]
+    tm.assert_frame_equal(result, expected)
+
+
+def test_concat_preserve_column_order_differing_columns():
+    # GH 4588 regression test
+    # for new columns in concat
+    dfa = pd.DataFrame(columns=['C', 'A'], data=[[1, 2]])
+    dfb = pd.DataFrame(columns=['C', 'Z'], data=[[5, 6]])
+    result = pd.concat([dfa, dfb], ignore_index=True, sort=True)
+
+    expected = pd.DataFrame({"A": [2, None], "C": [1, 5],
+                             "Z": [None, 6]}, columns=["A", "C", "Z"])
+    tm.assert_frame_equal(result, expected)
+
+
+def test_concat_preserve_column_order_uneven_data():
+    # GH 4588 regression test
+    # add to column, concat with uneven data
+    df = pd.DataFrame()
+    df['b'] = [1, 2, 3]
+    df['c'] = [1, 2, 3]
+    df['a'] = [1, 2, 3]
+    df2 = pd.DataFrame({'a': [4, 5]})
+    result = pd.concat([df, df2], sort=True)
+    expected = pd.DataFrame({
+        'a': [1, 2, 3, 4, 5],
+        'b': [1, 2, 3, None, None],
+        'c': [1, 2, 3, None, None]
+    }, index=[0, 1, 2, 0, 1])
+    tm.assert_frame_equal(result, expected)
+
+
+def test_concat_aligned_sort():
+    # GH-4588
+    df = pd.DataFrame({"c": [1, 2], "b": [3, 4], 'a': [5, 6]},
+                      columns=['c', 'b', 'a'])
+    result = pd.concat([df, df], sort=True, ignore_index=True)
+    expected = pd.DataFrame({'a': [5, 6, 5, 6], 'b': [3, 4, 3, 4],
+                             'c': [1, 2, 1, 2]},
+                            columns=['a', 'b', 'c'])
+    tm.assert_frame_equal(result, expected)
+
```

Comment:
not that I don't mind more tests, but are any of these duplicates of existing ones?

Reply:
Yeah, probably. I'll try to de-duplicate a bit.